lib/rpmsg/rpmsg_virtio.c

```diff
@@ -147,6 +147,7 @@ static void *rpmsg_virtio_get_tx_buffer(struct rpmsg_virtio_device *rvdev,
 			data = rpmsg_virtio_shm_pool_get_buffer(rvdev->shpool,
```

Comment:
@xiaoxiang781216 @arnopo 

In the case of master, the function rpmsg_virtio_get_tx_buffer keeps carving out a new buffer from shared memory if no buffer is found in the used-ring. If the remote receiver continuously holds buffers on the other side, the master will keep creating new buffers. I don't see a protection mechanism placed anywhere in the code which disallows more buffer creation than the available buffer descriptors. I have produced this condition quite easily by holding buffers at the remote side indefinitely. Eventually, when buffers are released by the remote, the indexes all went crazy which leads to a segment fault.

From remote to master communication there is no problem because all buffers are created by the master for remote at the time of rpmsg device initialization. And the buffers remain the same as the number of descriptors.

I don't understand the reason, why for Master the code workflow of buffer creation is different? Ideally for master also, all buffers should be created at the time of initialization. Then there is no need to create new buffers inside this function.

Reply:
> @xiaoxiang781216 @arnopo
> 
> In the case of master, the function rpmsg_virtio_get_tx_buffer keeps carving out a new buffer from shared memory if no buffer is found in the used-ring. If the remote receiver continuously holds buffers on the other side, the master will keep creating new buffers. I don't see a protection mechanism placed anywhere in the code which disallows more buffer creation than the available buffer descriptors. I have produced this condition quite easily by holding buffers at the remote side indefinitely. Eventually, when buffers are released by the remote, the indexes all went crazy which leads to a segment fault.
> 

Yes, this is an issue, I will try to find a method to fix it. Thanks for report.

> From remote to master communication there is no problem because all buffers are created by the master for remote at the time of rpmsg device initialization. And the buffers remain the same as the number of descriptors.
> 

> I don't understand the reason, why for Master the code workflow of buffer creation is different? Ideally for master also, all buffers should be created at the time of initialization. Then there is no need to create new buffers inside this function.

Yes, the master could allocate all buffers in the initialization, but master has to manage these buffer with link list manually, I guss this is one of reason why the allocation delay to rpmsg_virtio_get_tx_buffer.